commands/migratelxc.go

```diff
@@ -0,0 +1,291 @@
+	return &cmd.Info{
+		Name:    "migrate-lxc",
+		Args:    "<environment name>",
+		Purpose: "migrate of all the LXC containers in the specified environment to LXD",
+		Doc:     migrateLXCDoc,
+	}
+}
+
+func (c *migrateLXCCommand) Init(args []string) error {
+	args, err := c.baseClientCommand.init(args)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return cmd.CheckEmpty(args)
+}
+
+var migrateLXCImplDoc = `
+
+migrate-lxc-impl must be executed on an API server machine of a 1.25
+environment.
+
+The command will migrate all LXC containers in the environment to LXD.
+
+`
+
+func newMigrateLXCImplCommand() cmd.Command {
+	return &migrateLXCImplCommand{}
+}
+
+type migrateLXCImplCommand struct {
+	baseRemoteCommand
+	containerName string
+}
+
+func (c *migrateLXCImplCommand) Info() *cmd.Info {
+	return &cmd.Info{
+		Name:    "migrate-lxc-impl",
+		Purpose: "controller aspect of migrate-lxc",
+		Doc:     migrateLXCImplDoc,
+	}
+}
+
+func (c *migrateLXCImplCommand) Run(ctx *cmd.Context) error {
+	st, err := c.getState(ctx)
+	if err != nil {
+		return errors.Annotate(err, "getting state")
+	}
+	defer st.Close()
+
+	// Collect LXC container machines by host.
+	byHost, err := getLXCContainers(st)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// List LXD containers on each host, so we know which LXC
+	// containers still need migrating. This is required to make
+	// the migrate-lxc command idempotent/resumable.
+	var group errgroup.Group
+	hosts := make([]*state.Machine, 0, len(byHost))
+	lxdContainers := make([]map[string]*api.Container, len(byHost))
+	for host := range byHost {
+		i := len(hosts)
+		hosts = append(hosts, host)
+		host := host // copy for closure
+		group.Go(func() error {
+			containers, err := ListLXDContainers(host)
+			if err != nil {
+				return errors.Annotatef(err,
+					"listing LXD containers for host %q",
+					host.Id(),
+				)
+			}
+			lxdContainers[i] = containers
+			return nil
+		})
+	}
+	if err := group.Wait(); err != nil {
+		return errors.Annotate(err, "listing LXD containers")
+	}
+	lxdByHost := make(map[*state.Machine]map[string]*api.Container)
+	for i, containers := range lxdContainers {
+		lxdByHost[hosts[i]] = containers
+	}
+
+	// Filter out the LXC containers that have already been migrated.
+	environUUID := st.EnvironUUID()
+	namespace, err := instance.NewNamespace(environUUID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	oldNames := make(map[*state.Machine]string) // lxc container -> old name
+	newNames := make(map[string]string)         // old name -> new name
+	for host, containers := range byHost {
+		lxdContainers := lxdByHost[host]
+		nonMigrated := make([]*state.Machine, 0, len(containers))
+		for _, container := range containers {
+			instanceId, err := container.InstanceId()
+			if err != nil {
+				return err
+			}
+			oldName := string(instanceId)
+			oldNames[container] = oldName
+			newName, err := namespace.Hostname(container.Id())
```

Comment:
So `namespace.Hostname(id)` will tell you what the name of the container would be in Juju2, is that right?